File: golf/intel.kt
import java.math.BigInteger
import java.net.InetAddress
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.file.Files
import java.nio.file.Paths
import java.util.Locale
import kotlin.math.ln
import kotlin.math.min
import kotlin.math.round

val FLAGS = listOf("vpn","proxy","tor","malware","c2","scanner","brute_force","spammer",
    "compromised","datacenter","cdn","anycast","crawler","bot","cloud","private_relay",
    "anonymizer","mobile","isp","government")
val SEV = doubleArrayOf(30.0,25.0,45.0,95.0,95.0,55.0,70.0,65.0,75.0,15.0,
    5.0,0.0,10.0,40.0,10.0,15.0,35.0,0.0,0.0,0.0)
val LEVELS = listOf(80.0 to "critical", 60.0 to "high", 35.0 to "medium", 15.0 to "low")

data class Match(val source: String, val provider: String, val range: String,
                 val flags: List<String>, val weight: Double)

class DB(
    val v4s: LongArray, val v4e: LongArray, val v4v: IntArray, val v4m: LongArray,
    val v6s: Array<BigInteger>, val v6e: Array<BigInteger>, val v6v: IntArray,
    val v6m: Array<BigInteger>, val vt: Array<LongArray>, val strings: Array<String>,
    val weights: DoubleArray
)

fun u64BI(v: Long): BigInteger {
    var r = BigInteger.valueOf(v and 0x7FFFFFFFFFFFFFFFL)
    if (v < 0) r = r.setBit(63)
    return r
}

fun load(path: String): DB {
    val data = Files.readAllBytes(Paths.get(path))
    val b = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
    if (b.getInt(0) != 6) error("unsupported version")
    val h = LongArray(19) { b.getLong(8 + it * 8) }
    val cn = h[0].toInt(); val ln = h[1].toInt(); val v6n = h[2].toInt()
    val valn = h[3].toInt(); val strn = h[4].toInt()
    val oBucket = h[5].toInt(); val oSLo = h[6].toInt(); val oLens = h[7].toInt()
    val oVals = h[8].toInt(); val oLs = h[9].toInt(); val oLe = h[10].toInt()
    val oLv = h[11].toInt(); val oV6s = h[12].toInt(); val oV6e = h[13].toInt()
    val oV6v = h[14].toInt(); val oVt = h[15].toInt(); val oSi = h[16].toInt()
    val oSd = h[17].toInt()

    val bi = LongArray(65537) { (b.getInt(oBucket + it * 4).toLong()) and 0xFFFFFFFFL }
    val N = cn + ln
    val starts = LongArray(N); val ends = LongArray(N); val vals = IntArray(N)
    var j = 0L
    for (bk in 0 until 65536) {
        while (j < bi[bk + 1]) {
            val idx = j.toInt()
            val lo = (b.getShort(oSLo + idx * 2).toInt() and 0xFFFF).toLong()
            val s = (bk.toLong() shl 16) or lo
            starts[idx] = s
            ends[idx] = s + (b.getShort(oLens + idx * 2).toInt() and 0xFFFF)
            vals[idx] = b.getShort(oVals + idx * 2).toInt() and 0xFFFF
            j++
        }
    }
    for (i in 0 until ln) {
        starts[cn + i] = (b.getInt(oLs + i * 4).toLong()) and 0xFFFFFFFFL
        ends[cn + i] = (b.getInt(oLe + i * 4).toLong()) and 0xFFFFFFFFL
        vals[cn + i] = b.getShort(oLv + i * 2).toInt() and 0xFFFF
    }

    val order = (0 until N).sortedBy { starts[it] }
    val s2 = LongArray(N) { starts[order[it]] }
    val e2 = LongArray(N) { ends[order[it]] }
    val v2 = IntArray(N) { vals[order[it]] }
    val m4 = LongArray(N)
    var mx = 0L
    for (i in 0 until N) { if (e2[i] > mx) mx = e2[i]; m4[i] = mx }

    val v6s = Array(v6n) {
        u64BI(b.getLong(oV6s + it * 16 + 8)).shiftLeft(64)
            .or(u64BI(b.getLong(oV6s + it * 16)))
    }
    val v6e = Array(v6n) {
        u64BI(b.getLong(oV6e + it * 16 + 8)).shiftLeft(64)
            .or(u64BI(b.getLong(oV6e + it * 16)))
    }
    val v6v = IntArray(v6n) { b.getShort(oV6v + it * 2).toInt() and 0xFFFF }
    val v6m = Array(v6n) { BigInteger.ZERO }
    var m6 = BigInteger.ZERO
    for (i in 0 until v6n) { if (v6e[i] > m6) m6 = v6e[i]; v6m[i] = m6 }

    val vt = Array(valn) { i ->
        LongArray(4) { k -> (b.getInt(oVt + (i * 4 + k) * 4).toLong()) and 0xFFFFFFFFL }
    }
    val strings = Array(strn) {
        val so = (b.getInt(oSi + it * 8).toLong() and 0xFFFFFFFFL).toInt()
        val sl = (b.getInt(oSi + it * 8 + 4).toLong() and 0xFFFFFFFFL).toInt()
        String(data, oSd + so, sl, Charsets.UTF_8)
    }

    val weights = DoubleArray(20)
    if (N > 0) {
        val cnt = IntArray(20)
        for (vid in v2) {
            val bits = vt[vid][0]
            for (i in 0 until 20) if ((bits and (1L shl i)) != 0L) cnt[i]++
        }
        for (i in 0 until 20) {
            val cc = if (cnt[i] == 0) 1 else cnt[i]
            weights[i] = SEV[i] * (1 + ln(N.toDouble() / cc) / ln(2.0) / 24)
        }
    } else SEV.copyInto(weights)

    return DB(s2, e2, v2, m4, v6s, v6e, v6v, v6m, vt, strings, weights)
}

fun r1(x: Double) = round(x * 10) / 10.0

fun v4str(ip: Long) = "${(ip shr 24) and 0xFF}.${(ip shr 16) and 0xFF}." +
    "${(ip shr 8) and 0xFF}.${ip and 0xFF}"

fun v6str(ip: BigInteger): String {
    val full = ByteArray(16)
    val raw = ip.toByteArray()
    val copy = min(raw.size, 16)
    System.arraycopy(raw, raw.size - copy, full, 16 - copy, copy)
    return InetAddress.getByAddress(full).hostAddress
}

fun buildMatch(db: DB, vid: Int, range: String): Match {
    val bits = db.vt[vid][0]
    val fl = mutableListOf<String>()
    var mxw = 0.0
    for (i in 0 until 20) {
        if ((bits and (1L shl i)) != 0L) {
            fl.add(FLAGS[i])
            if (db.weights[i] > mxw) mxw = db.weights[i]
        }
    }
    return Match(db.strings[db.vt[vid][2].toInt()],
        db.strings[db.vt[vid][1].toInt()], range, fl, r1(mxw))
}

fun upperBoundL(arr: LongArray, key: Long): Int {
    var lo = 0; var hi = arr.size
    while (lo < hi) {
        val mid = (lo + hi) ushr 1
        if (arr[mid] <= key) lo = mid + 1 else hi = mid
    }
    return lo
}

fun upperBoundB(arr: Array<BigInteger>, key: BigInteger): Int {
    var lo = 0; var hi = arr.size
    while (lo < hi) {
        val mid = (lo + hi) ushr 1
        if (arr[mid] <= key) lo = mid + 1 else hi = mid
    }
    return lo
}

fun lookup(db: DB, ipStr: String): String {
    val addr = InetAddress.getByName(ipStr)
    val raw = addr.address
    val matches = mutableListOf<Match>()
    if (raw.size == 4) {
        var ip = 0L
        for (v in raw) ip = (ip shl 8) or (v.toLong() and 0xFF)
        var i = upperBoundL(db.v4s, ip)
        while (i > 0) {
            i--
            if (db.v4m[i] < ip) break
            if (db.v4e[i] >= ip) matches.add(buildMatch(db, db.v4v[i],
                "${v4str(db.v4s[i])}-${v4str(db.v4e[i])}"))
        }
    } else {
        val ip = BigInteger(1, raw)
        var i = upperBoundB(db.v6s, ip)
        while (i > 0) {
            i--
            if (db.v6m[i] < ip) break
            if (db.v6e[i] >= ip) matches.add(buildMatch(db, db.v6v[i],
                "${v6str(db.v6s[i])}-${v6str(db.v6e[i])}"))
        }
    }
    matches.sortByDescending { it.weight }

    val allFlags = LinkedHashSet<String>()
    val providers = LinkedHashSet<String>()
    val srcs = LinkedHashSet<String>()
    val rankedSet = LinkedHashSet<String>()
    for (m in matches) {
        for (f in m.flags) { allFlags.add(f); rankedSet.add(f) }
        if (m.provider.isNotEmpty()) providers.add(m.provider)
        srcs.add("${m.provider}|${m.source}")
    }
    val flagIdx = FLAGS.withIndex().associate { (i, f) -> f to i }
    val ranked = rankedSet.sortedByDescending { db.weights[flagIdx[it]!!] }

    var score = 0.0
    if (ranked.isNotEmpty()) {
        val top = db.weights[flagIdx[ranked[0]]!!]
        val ex = ranked.drop(1).sumOf { db.weights[flagIdx[it]!!] }
        score = r1(min(100.0, (top + ex * 0.15) *
            (1 + 0.08 * ln(srcs.size + 1.0) / ln(2.0))))
    }
    var verdict = "clean"
    if (matches.isNotEmpty()) {
        verdict = "minimal"
        for ((t, name) in LEVELS) if (score >= t) { verdict = name; break }
    }

    val provList = providers.toMutableList()
    if (provList.any { it.equals("tor", true) }) {
        provList.removeAll { it.equals("tor", true) }
        provList.add(0, "Tor")
    }
    val reasons = ranked.take(5)
    val topProv = provList.firstOrNull() ?: ""

    return buildString {
        append("{\n")
        append("  \"ip\": ").append(jstr(ipStr)).append(",\n")
        append("  \"found\": ").append(matches.isNotEmpty()).append(",\n")
        append("  \"verdict\": ").append(jstr(verdict)).append(",\n")
        append("  \"score\": ").append(num(score)).append(",\n")
        append("  \"detections\": ").append(matches.size).append(",\n")
        append("  \"sources\": ").append(srcs.size).append(",\n")
        append("  \"top_provider\": ").append(jstr(topProv)).append(",\n")
        append("  \"providers\": ").append(jarr(provList, "  ")).append(",\n")
        append("  \"flags\": ").append(jarr(allFlags.toList(), "  ")).append(",\n")
        append("  \"reasons\": ").append(jarr(reasons, "  ")).append(",\n")
        append("  \"matches\": ").append(jmatches(matches, "  ")).append("\n")
        append("}")
    }
}

fun num(d: Double): String =
    if (d == d.toLong().toDouble()) "${d.toLong()}.0"
    else "%.1f".format(Locale.ROOT, d)

fun jstr(s: String): String {
    val b = StringBuilder("\"")
    for (c in s) when (c.code) {
        0x22 -> b.append("\\\"")
        0x5C -> b.append("\\\\")
        0x0A -> b.append("\\n")
        0x0D -> b.append("\\r")
        0x09 -> b.append("\\t")
        0x08 -> b.append("\\b")
        0x0C -> b.append("\\f")
        else -> if (c.code < 0x20) b.append("\\u%04x".format(c.code)) else b.append(c)
    }
    return b.append("\"").toString()
}

fun jarr(items: List<String>, indent: String): String {
    if (items.isEmpty()) return "[]"
    val inner = "$indent  "
    return items.joinToString(",\n", "[\n", "\n$indent]") { "$inner${jstr(it)}" }
}

fun jmatches(ms: List<Match>, indent: String): String {
    if (ms.isEmpty()) return "[]"
    val inner = "$indent  "
    val deep = "$inner  "
    return ms.joinToString(",\n", "[\n", "\n$indent]") { m ->
        "$inner{\n" +
        "$deep\"source\": ${jstr(m.source)},\n" +
        "$deep\"provider\": ${jstr(m.provider)},\n" +
        "$deep\"range\": ${jstr(m.range)},\n" +
        "$deep\"flags\": ${jarr(m.flags, deep)},\n" +
        "$deep\"weight\": ${num(m.weight)}\n" +
        "$inner}"
    }
}

fun main(args: Array<String>) {
    val ip = if (args.isNotEmpty()) args[0] else "8.8.8.8"
    val db = load("../intel.bin")
    println(lookup(db, ip))
}

File: golf/intel.swift
import Foundation
#if canImport(Darwin)
import Darwin
#elseif canImport(Glibc)
import Glibc
#endif

let FLAGS = ["vpn","proxy","tor","malware","c2","scanner","brute_force","spammer",
    "compromised","datacenter","cdn","anycast","crawler","bot","cloud","private_relay",
    "anonymizer","mobile","isp","government"]
let SEV: [Double] = [30,25,45,95,95,55,70,65,75,15,5,0,10,40,10,15,35,0,0,0]
let LEVELS: [(Double, String)] = [(80,"critical"),(60,"high"),(35,"medium"),(15,"low")]

struct U128: Comparable {
    let hi: UInt64; let lo: UInt64
    static func < (a: U128, b: U128) -> Bool {
        a.hi != b.hi ? a.hi < b.hi : a.lo < b.lo
    }
    static func == (a: U128, b: U128) -> Bool { a.hi == b.hi && a.lo == b.lo }
}

struct Match {
    let source: String; let provider: String; let range: String
    let flags: [String]; let weight: Double
}

final class DB {
    var v4s: [UInt32] = [], v4e: [UInt32] = [], v4m: [UInt32] = [], v4v: [UInt16] = []
    var v6s: [U128] = [], v6e: [U128] = [], v6m: [U128] = [], v6v: [UInt16] = []
    var vt: [[UInt32]] = [], strings: [String] = []
    var weights = [Double](repeating: 0, count: 20)
}

func readU16(_ d: Data, _ o: Int) -> UInt16 {
    d.withUnsafeBytes { $0.loadUnaligned(fromByteOffset: o, as: UInt16.self) }
}
func readU32(_ d: Data, _ o: Int) -> UInt32 {
    d.withUnsafeBytes { $0.loadUnaligned(fromByteOffset: o, as: UInt32.self) }
}
func readU64(_ d: Data, _ o: Int) -> UInt64 {
    d.withUnsafeBytes { $0.loadUnaligned(fromByteOffset: o, as: UInt64.self) }
}

func load(_ path: String) -> DB {
    let d = try! Data(contentsOf: URL(fileURLWithPath: path))
    guard readU32(d, 0) == 6 else { fatalError("unsupported version") }
    let h = (0..<19).map { Int(readU64(d, 8 + $0 * 8)) }
    let cn = h[0], ln = h[1], v6n = h[2], valn = h[3], strn = h[4]
    let off = Array(h[5...])
    let bi = (0...65536).map { readU32(d, off[0] + $0 * 4) }
    let N = cn + ln
    var starts = [UInt32](repeating: 0, count: N)
    var ends = [UInt32](repeating: 0, count: N)
    var vals = [UInt16](repeating: 0, count: N)
    for b in 0..<65536 {
        for j in Int(bi[b])..<Int(bi[b+1]) {
            let lo = UInt32(readU16(d, off[1] + j * 2))
            let s = (UInt32(b) << 16) | lo
            starts[j] = s
            ends[j] = s + UInt32(readU16(d, off[2] + j * 2))
            vals[j] = readU16(d, off[3] + j * 2)
        }
    }
    for i in 0..<ln {
        starts[cn + i] = readU32(d, off[4] + i * 4)
        ends[cn + i] = readU32(d, off[5] + i * 4)
        vals[cn + i] = readU16(d, off[6] + i * 2)
    }
    let order = (0..<N).sorted { starts[$0] < starts[$1] }
    let db = DB()
    db.v4s = order.map { starts[$0] }
    db.v4e = order.map { ends[$0] }
    db.v4v = order.map { vals[$0] }
    db.v4m = [UInt32](repeating: 0, count: N)
    var mx: UInt32 = 0
    for i in 0..<N { if db.v4e[i] > mx { mx = db.v4e[i] }; db.v4m[i] = mx }

    db.v6s = (0..<v6n).map {
        U128(hi: readU64(d, off[7] + $0 * 16 + 8), lo: readU64(d, off[7] + $0 * 16))
    }
    db.v6e = (0..<v6n).map {
        U128(hi: readU64(d, off[8] + $0 * 16 + 8), lo: readU64(d, off[8] + $0 * 16))
    }
    db.v6v = (0..<v6n).map { readU16(d, off[9] + $0 * 2) }
    db.v6m = [U128](repeating: U128(hi: 0, lo: 0), count: v6n)
    var m6 = U128(hi: 0, lo: 0)
    for i in 0..<v6n { if db.v6e[i] > m6 { m6 = db.v6e[i] }; db.v6m[i] = m6 }

    db.vt = (0..<valn).map { i in
        (0..<4).map { k in readU32(d, off[10] + (i * 4 + k) * 4) }
    }
    let sd = off[12]
    db.strings = (0..<strn).map { i in
        let so = Int(readU32(d, off[11] + i * 8))
        let sl = Int(readU32(d, off[11] + i * 8 + 4))
        return String(data: d.subdata(in: (sd+so)..<(sd+so+sl)), encoding: .utf8) ?? ""
    }
    if N > 0 {
        var cnt = [Int](repeating: 0, count: 20)
        for vid in db.v4v {
            let bits = db.vt[Int(vid)][0]
            for i in 0..<20 { if bits & (UInt32(1) << i) != 0 { cnt[i] += 1 } }
        }
        for i in 0..<20 {
            let c = max(cnt[i], 1)
            db.weights[i] = SEV[i] * (1 + log2(Double(N) / Double(c)) / 24)
        }
    } else {
        db.weights = SEV
    }
    return db
}

func r1(_ x: Double) -> Double { (x * 10).rounded() / 10 }

func parseIP(_ s: String) -> (v4: UInt32?, v6: U128?) {
    var buf4 = in_addr()
    if inet_pton(AF_INET, s, &buf4) == 1 {
        return (UInt32(bigEndian: buf4.s_addr), nil)
    }
    var buf6 = in6_addr()
    if inet_pton(AF_INET6, s, &buf6) == 1 {
        var bytes = [UInt8](repeating: 0, count: 16)
        withUnsafeBytes(of: &buf6) { raw in
            for i in 0..<16 { bytes[i] = raw[i] }
        }
        var hi: UInt64 = 0, lo: UInt64 = 0
        for i in 0..<8 { hi = (hi << 8) | UInt64(bytes[i]) }
        for i in 8..<16 { lo = (lo << 8) | UInt64(bytes[i]) }
        return (nil, U128(hi: hi, lo: lo))
    }
    fatalError("invalid ip: \(s)")
}

func fmtV4(_ ip: UInt32) -> String {
    "\((ip >> 24) & 0xFF).\((ip >> 16) & 0xFF).\((ip >> 8) & 0xFF).\(ip & 0xFF)"
}

func fmtV6(_ ip: U128) -> String {
    var bytes = [UInt8](repeating: 0, count: 16)
    for i in 0..<8 { bytes[i] = UInt8((ip.hi >> (56 - i*8)) & 0xFF) }
    for i in 0..<8 { bytes[8+i] = UInt8((ip.lo >> (56 - i*8)) & 0xFF) }
    var addr = in6_addr()
    withUnsafeMutableBytes(of: &addr) { raw in
        for i in 0..<16 { raw[i] = bytes[i] }
    }
    var out = [CChar](repeating: 0, count: Int(INET6_ADDRSTRLEN))
    inet_ntop(AF_INET6, &addr, &out, socklen_t(INET6_ADDRSTRLEN))
    return String(cString: out)
}

func upperU32(_ a: [UInt32], _ k: UInt32) -> Int {
    var lo = 0, hi = a.count
    while lo < hi { let m = (lo + hi) / 2; if a[m] > k { hi = m } else { lo = m + 1 } }
    return lo
}

func upperU128(_ a: [U128], _ k: U128) -> Int {
    var lo = 0, hi = a.count
    while lo < hi { let m = (lo + hi) / 2; if a[m] > k { hi = m } else { lo = m + 1 } }
    return lo
}

func buildMatch(_ db: DB, _ vid: UInt16, _ rng: String) -> Match {
    let row = db.vt[Int(vid)]
    let bits = row[0]
    var fl: [String] = []
    var mxw: Double = 0
    for i in 0..<20 {
        if bits & (UInt32(1) << i) != 0 {
            fl.append(FLAGS[i])
            if db.weights[i] > mxw { mxw = db.weights[i] }
        }
    }
    return Match(source: db.strings[Int(row[2])], provider: db.strings[Int(row[1])],
                 range: rng, flags: fl, weight: r1(mxw))
}

func num(_ d: Double) -> String {
    if d == d.rounded() && abs(d) < 1e16 {
        return "\(Int64(d)).0"
    }
    return String(format: "%.1f", d)
}

func jstr(_ s: String) -> String {
    var b = "\""
    for c in s.unicodeScalars {
        switch c.value {
        case 0x22: b += "\\\""
        case 0x5C: b += "\\\\"
        case 0x0A: b += "\\n"
        case 0x0D: b += "\\r"
        case 0x09: b += "\\t"
        case 0x08: b += "\\b"
        case 0x0C: b += "\\f"
        default:
            if c.value < 0x20 { b += String(format: "\\u%04x", c.value) }
            else { b += String(c) }
        }
    }
    return b + "\""
}

func jarr(_ items: [String], _ indent: String) -> String {
    if items.isEmpty { return "[]" }
    let inner = indent + "  "
    return "[\n" + items.map { "\(inner)\(jstr($0))" }.joined(separator: ",\n")
        + "\n\(indent)]"
}

func jmatches(_ ms: [Match], _ indent: String) -> String {
    if ms.isEmpty { return "[]" }
    let inner = indent + "  "
    let deep = inner + "  "
    let parts = ms.map { m -> String in
        "\(inner){\n"
        + "\(deep)\"source\": \(jstr(m.source)),\n"
        + "\(deep)\"provider\": \(jstr(m.provider)),\n"
        + "\(deep)\"range\": \(jstr(m.range)),\n"
        + "\(deep)\"flags\": \(jarr(m.flags, deep)),\n"
        + "\(deep)\"weight\": \(num(m.weight))\n"
        + "\(inner)}"
    }
    return "[\n" + parts.joined(separator: ",\n") + "\n\(indent)]"
}

func lookup(_ db: DB, _ ipStr: String) -> String {
    let (v4, v6) = parseIP(ipStr)
    var matches: [Match] = []
    if let ip = v4 {
        var i = upperU32(db.v4s, ip)
        while i > 0 {
            i -= 1
            if db.v4m[i] < ip { break }
            if db.v4e[i] >= ip {
                matches.append(buildMatch(db, db.v4v[i],
                    "\(fmtV4(db.v4s[i]))-\(fmtV4(db.v4e[i]))"))
            }
        }
    } else if let ip = v6 {
        var i = upperU128(db.v6s, ip)
        while i > 0 {
            i -= 1
            if db.v6m[i] < ip { break }
            if db.v6e[i] >= ip {
                matches.append(buildMatch(db, db.v6v[i],
                    "\(fmtV6(db.v6s[i]))-\(fmtV6(db.v6e[i]))"))
            }
        }
    }
    matches.sort { $0.weight > $1.weight }

    var allFlags: [String] = [], providers: [String] = [], srcs: [String] = []
    var rankedSet: [String] = []
    var seenFlag = Set<String>(), seenProv = Set<String>(), seenSrc = Set<String>()
    for m in matches {
        for f in m.flags where !seenFlag.contains(f) {
            seenFlag.insert(f); allFlags.append(f); rankedSet.append(f)
        }
        if !m.provider.isEmpty && !seenProv.contains(m.provider) {
            seenProv.insert(m.provider); providers.append(m.provider)
        }
        let key = "\(m.provider)|\(m.source)"
        if !seenSrc.contains(key) { seenSrc.insert(key); srcs.append(key) }
    }
    let fi = Dictionary(uniqueKeysWithValues: FLAGS.enumerated().map { ($1, $0) })
    let ranked = rankedSet.sorted { db.weights[fi[$0]!] > db.weights[fi[$1]!] }

    var score = 0.0
    if !ranked.isEmpty {
        let top = db.weights[fi[ranked[0]]!]
        let ex = ranked.dropFirst().reduce(0.0) { $0 + db.weights[fi[$1]!] }
        score = r1(min(100, (top + ex * 0.15) * (1 + 0.08 * log2(Double(srcs.count + 1)))))
    }
    var verdict = "clean"
    if !matches.isEmpty {
        verdict = "minimal"
        for (t, n) in LEVELS where score >= t { verdict = n; break }
    }

    var provList = providers
    if let idx = provList.firstIndex(where: { $0.lowercased() == "tor" }) {
        provList.remove(at: idx)
        provList.insert("Tor", at: 0)
    }
    let reasons = Array(ranked.prefix(5))
    let topProv = provList.first ?? ""

    var out = "{\n"
    out += "  \"ip\": \(jstr(ipStr)),\n"
    out += "  \"found\": \(matches.isEmpty ? "false" : "true"),\n"
    out += "  \"verdict\": \(jstr(verdict)),\n"
    out += "  \"score\": \(num(score)),\n"
    out += "  \"detections\": \(matches.count),\n"
    out += "  \"sources\": \(srcs.count),\n"
    out += "  \"top_provider\": \(jstr(topProv)),\n"
    out += "  \"providers\": \(jarr(provList, "  ")),\n"
    out += "  \"flags\": \(jarr(allFlags, "  ")),\n"
    out += "  \"reasons\": \(jarr(reasons, "  ")),\n"
    out += "  \"matches\": \(jmatches(matches, "  "))\n"
    out += "}"
    return out
}

let args = CommandLine.arguments
let ip = args.count > 1 ? args[1] : "8.8.8.8"
let db = load("../intel.bin")
print(lookup(db, ip))
